File: data_mining.cpp
#include <bits/stdc++.h>
#define NO_OF_SAMPLE 100000
#define DATA_FILE_NAME "data_set_aminar_dblp.v11/dblp_papers_v11.txt"
//#define DATA_FILE_NAME "data_set_aminar_dblp.v11/sample_of_dblp_papers_v11.txt"

using namespace std;

class Authors
{
	public :
		string name;
		string id;
		string org;

		Authors();
};

class Venue
{
	public :
		string id;
		string raw;

		Venue();
};

class Fos
{
	public :
		string name;
		double w;

		Fos();
};

class Word_index
{
	public :
		string word;
		vector<string> index;

		Word_index();
};

class Indexed_abstract
{
	public :
		int IndexLength;
		vector<Word_index> InvertedIndex;

		Indexed_abstract();
};

class Paper
{
	public :
		string id;
		string title;
		list<Authors> authors;
		Venue venue;
		int year;
		list<string> keywords;
		list<Fos> fos;
		list<string> references;
		int n_citation;
		string page_start;
		string page_end;
		string doc_type;
		string lang;
		string publisher;
		string volume;
		string issue;
		string issn;
		string isbn;
		string doi;
		string pdf;
		list<string> url;
		string abstract;
		Indexed_abstract indexed_abstract;

		Paper();
};

class Collection
{
	public :
		int collection_size;
		vector<Paper> paper;
		map<string, int> attributes;

		Collection();
		void read_data();
		void print_data();
		void map_attributes_to_integer();
};

string get_next_word(string this_paper, int *j)
{
	string this_word;
	while(((*j) < this_paper.size()) && (this_paper[(*j)] != '"'))
	{
		(*j)++;
	}
	(*j)++;
	while(((*j) < this_paper.size()) && (this_paper[(*j)] != '"'))
	{
		this_word.push_back(this_paper[(*j)]);
		(*j)++;	
	}
	(*j)++;
	return this_word;
}

ifstream input_file;

int main()
{
	input_file.open(DATA_FILE_NAME);
	if(!input_file)
	{
		cout<<"Error!, Couldn't able to open the file."<<endl;
		return 0;
	}
	Collection collection;
	collection.read_data();
	//collection.print_data();
	input_file.close();
	return 0;
}

Authors :: Authors()
{
	name = "";
	id = "";
	org = "";
}

Venue :: Venue()
{
	id = "";
	raw = "";
}

Word_index :: Word_index()
{
	word = "";
	index.resize(0);
}

Fos :: Fos()
{
	name = "";
	w = 0.0;
}

Indexed_abstract :: Indexed_abstract()
{
	IndexLength = 0;
	InvertedIndex.resize(0);
}

Paper :: Paper()
{
	id = "";
	title = "";
	authors.resize(0);
	year = 0;
	keywords.resize(0);
	fos.resize(0);
	references.resize(0);
	n_citation = 0;
	page_start = "";
	page_end = "";
	doc_type = "";
	lang = "";
	publisher = "";
	volume = "";
	issue = "";
	issn = "";
	isbn = "";
	doi = "";
	pdf = "";
	url.resize(0);
	abstract = "";
}

Collection :: Collection()
{
	collection_size = NO_OF_SAMPLE;
	paper.resize(NO_OF_SAMPLE);
	map_attributes_to_integer();
}

void Collection :: print_data()
{
	for(int i = 0; i < collection_size; i++)
	{
		cout<<"id : "<<paper[i].id<<endl;
		cout<<"title : "<<paper[i].title<<endl;
		cout<<"authors : "<<endl;
		for(auto x : paper[i].authors)
		{
			cout<<'\t'<<"name : "<<x.name;
			cout<<'\t'<<"id : "<<x.id;
			cout<<'\t'<<"org : "<<x.org<<endl;
		}
		cout<<"keywords : "<<endl;
		for(auto x : paper[i].keywords)
		{
			cout<<'\t'<<x<<endl;
		}
		cout<<"fos : "<<endl;
		for(auto x : paper[i].fos)
		{
			cout<<'\t'<<"name : "<<x.name;
			cout<<'\t'<<"w : "<<x.w<<endl;
		}
		cout<<"references : "<<endl;
		for(auto x : paper[i].references)
		{
			cout<<'\t'<<x<<endl;
		}
		cout<<"venue : "<<endl;
		cout<<'\t'<<"raw : "<<paper[i].venue.raw<<endl;
		cout<<'\t'<<"id : "<<paper[i].venue.id<<endl;
		cout<<"year : "<<paper[i].year<<endl;
		cout<<"n_citation : "<<paper[i].n_citation<<endl;
		cout<<"page_start : "<<paper[i].page_start<<endl;
		cout<<"page_end : "<<paper[i].page_end<<endl;
		cout<<"doc_type : "<<paper[i].doc_type<<endl;
		cout<<"lang : "<<paper[i].lang<<endl;
		cout<<"publisher : "<<paper[i].publisher<<endl;
		cout<<"volume : "<<paper[i].volume<<endl;
		cout<<"issue : "<<paper[i].issue<<endl;
		cout<<"issn : "<<paper[i].issn<<endl;
		cout<<"isbn : "<<paper[i].isbn<<endl;
		cout<<"doi : "<<paper[i].doi<<endl;
		cout<<"pdf : "<<paper[i].pdf<<endl;
		cout<<"url : "<<endl;
		for(auto x : paper[i].url)
		{
			cout<<'\t'<<x<<endl;
		}
		cout<<"indexed_abstract : "<<endl;
		cout<<'\t'<<"IndexLength : "<<paper[i].indexed_abstract.IndexLength<<endl;
		cout<<'\t'<<"InvertedIndex : "<<endl;
		for(auto x : paper[i].indexed_abstract.InvertedIndex)
		{
			cout<<"\t\t"<<"word : "<<x.word<<"\t\tIndex : ";
			for(auto y : x.index)
			{
				cout<<y<<",";
			}
			cout<<"\n";
		}
		cout<<endl;
	}
}

void Collection :: read_data()
{
	for(int i = 0; i < collection_size; i++)
	{
		string this_paper;
		getline(input_file, this_paper);
		if(this_paper == "")
		{
			cout<<"Only "<<i<<" papers are present in given file."<<endl;
			paper.resize(i);
			collection_size = i;
			break;
		}
		int j = 0;
		while(j < this_paper.size())
		{
			string this_word = get_next_word(this_paper, &j);
			auto it = attributes.find(this_word);
			if(it != attributes.end())
			{
				int key = it->second;
				switch(key)
				{
					case 0 : 	paper[i].id = get_next_word(this_paper, &j);
								break;

					case 1 :	paper[i].title = get_next_word(this_paper, &j);
								break;

					case 2 :	do
								{
									Authors temp;
									do
									{
										string temp1 = get_next_word(this_paper, &j);
										if(temp1 == "name")
										{
											temp.name = get_next_word(this_paper, &j);
										}
										else if(temp1 == "id")
										{
											temp.id = get_next_word(this_paper, &j);
										}
										else if(temp1 == "org")
										{
											temp.org = get_next_word(this_paper, &j);
										}
									}while((j < this_paper.size()) && (this_paper[j] != '}'));
									paper[i].authors.push_back(temp);
								}while((j < this_paper.size()) && (this_paper[j + 1] != ']'));
								break;

					case 3 :	do
								{
									string temp1 = get_next_word(this_paper, &j);
									if(temp1 == "id")
									{
										paper[i].venue.id = get_next_word(this_paper, &j);
									}
									else if(temp1 == "raw")
									{
										paper[i].venue.raw = get_next_word(this_paper, &j);
									}
								}while((j < this_paper.size()) && (this_paper[j] != '}'));
								break;

					case 4 :	j = j + 2;
								paper[i].year = 0;
								while((j < this_paper.size()) && (this_paper[j] >= '0') && (this_paper[j] <= '9'))
								{
									paper[i].year *= 10;
									paper[i].year += (this_paper[j] - '0');
									j++;
								}
								break;

					case 5 :	do
								{
									string temp1 = get_next_word(this_paper, &j);
									paper[i].keywords.push_back(temp1);
								}while((j < this_paper.size()) && (this_paper[j] != ']'));
								break;

					case 6 :	do
								{
									Fos temp;
									do
									{
										string temp1 = get_next_word(this_paper, &j);
										if(temp1 == "name")
										{
											temp.name = get_next_word(this_paper, &j);
										}
										else if(temp1 == "w")
										{
											j = j + 2;
											temp.w = 0;
											int flag = 0;
											while((j < this_paper.size()) && (((this_paper[j] >= '0') && (this_paper[j] <= '9')) || (this_paper[j] == '.')))
											{	
												if(this_paper[j] == '.')
												{
													flag = 1;
												}
												else if(flag == 0)
												{	
													temp.w *= 10;
													temp.w += (this_paper[j] - '0');
												}
												else if(flag >= 1)
												{
													temp.w += ((this_paper[j] - '0') / pow(10, flag));
													flag++;
												}
												j++;
											}
										}
									}while((j < this_paper.size()) && (this_paper[j] != '}'));
									paper[i].fos.push_back(temp);
								}while((j < this_paper.size()) && (this_paper[j + 1] != ']'));
								break;

					case 7 :	do
								{
									string temp1 = get_next_word(this_paper, &j);
									paper[i].references.push_back(temp1);
								}while((j < this_paper.size()) && (this_paper[j] != ']'));
								break;

					case 8 :	j = j + 2;
								paper[i].n_citation = 0;
								while((j < this_paper.size()) && (this_paper[j] >= '0') && (this_paper[j] <= '9'))
								{
									paper[i].n_citation *= 10;
									paper[i].n_citation += (this_paper[j] - '0');
									j++;
								}
								break;

					case 9 :	paper[i].page_start = get_next_word(this_paper, &j);
								break;

					case 10 :	paper[i].page_end = get_next_word(this_paper, &j);
								break;

					case 11 :	paper[i].doc_type = get_next_word(this_paper, &j);
								break;

					case 12 :	paper[i].lang = get_next_word(this_paper, &j);
								break;

					case 13 :	paper[i].publisher = get_next_word(this_paper, &j);
								break;

					case 14 :	paper[i].volume = get_next_word(this_paper, &j);
								break;

					case 15 :	paper[i].issue = get_next_word(this_paper, &j);
								break;

					case 16 :	paper[i].issn = get_next_word(this_paper, &j);
								break;

					case 17 :	paper[i].isbn = get_next_word(this_paper, &j);
								break;

					case 18 :	paper[i].doi = get_next_word(this_paper, &j);
								break;

					case 19 :	paper[i].pdf = get_next_word(this_paper, &j);
								break;

					case 20 :	do
								{
									string temp1 = get_next_word(this_paper, &j);
									paper[i].url.push_back(temp1);
								}while((j < this_paper.size()) && (this_paper[j] != ']'));
								break;

					case 21 :	paper[i].abstract = get_next_word(this_paper, &j);
								break;

					case 22 :	get_next_word(this_paper, &j);
								j = j + 2;
								paper[i].indexed_abstract.IndexLength = 0;
								while((j < this_paper.size()) && (this_paper[j] >= '0') && (this_paper[j] <= '9'))
								{
									paper[i].indexed_abstract.IndexLength *= 10;
									paper[i].indexed_abstract.IndexLength += (this_paper[j] - '0');
									j++;
								}
								get_next_word(this_paper, &j);
								do
								{
									Word_index temp;
									temp.word = get_next_word(this_paper, &j);
									j = j + 3;
									string temp1;
									do
									{
										if(this_paper[j] == ',')
										{
											temp.index.push_back(temp1);
											temp1 = "";
											j = j + 1;
										}
										else
										{
											temp1.push_back(this_paper[j]);
										}
										j = j + 1;
									}while((j < this_paper.size()) && (this_paper[j] != ']'));
									temp.index.push_back(temp1);
									paper[i].indexed_abstract.InvertedIndex.push_back(temp);
								}while((j < this_paper.size()) && (this_paper[j + 1] != '}'));
								break;

					default :
								break;
				}
			}
		}
	}
}

void Collection :: map_attributes_to_integer()
{
	attributes.insert(pair<string, int>("id" , 0));
	attributes.insert(pair<string, int>("title", 1));
	attributes.insert(pair<string, int>("authors", 2));
	attributes.insert(pair<string, int>("venue", 3));
	attributes.insert(pair<string, int>("year", 4));
	attributes.insert(pair<string, int>("keywords", 5));
	attributes.insert(pair<string, int>("fos", 6));
	attributes.insert(pair<string, int>("references", 7));
	attributes.insert(pair<string, int>("n_citation", 8));
	attributes.insert(pair<string, int>("page_start", 9));
	attributes.insert(pair<string, int>("page_end", 10));
	attributes.insert(pair<string, int>("doc_type", 11));
	attributes.insert(pair<string, int>("lang", 12));
	attributes.insert(pair<string, int>("publisher", 13));
	attributes.insert(pair<string, int>("volume", 14));
	attributes.insert(pair<string, int>("issue", 15));
	attributes.insert(pair<string, int>("issn", 16));
	attributes.insert(pair<string, int>("isbn", 17));
	attributes.insert(pair<string, int>("doi", 18));
	attributes.insert(pair<string, int>("pdf", 19));
	attributes.insert(pair<string, int>("url", 20));
	attributes.insert(pair<string, int>("abstract", 21));
	attributes.insert(pair<string, int>("indexed_abstract", 22));
}

File: data_mining.py
data_file_name = 'data_set_aminar_dblp.v11/dblp_papers_v11.txt'
#data_file_name = 'data_set_aminar_dblp.v11/sample_of_dblp_papers_v11.txt'
no_of_sample = 100000

def get_next_word(this_paper):
	global j
	this_word = ''
	while j < len(this_paper) and this_paper[j] != '"':
		j = j + 1
	j = j + 1
	while j < len(this_paper) and this_paper[j] != '"':
		this_word = this_word + this_paper[j]
		j = j + 1
	j = j + 1
	return this_word

class Authors:
	def __init__(self):
		self.name = ''
		self.identity = ''
		self.org = ''

class Venue:
	def __init__(self):
		self.identity = ''
		self.raw = ''

class Fos:
	def __init__(self):
		self.name = ''
		self.w = 0.0

class Indexed_abstract:
	def __init__(self):
		self.IndexLength = 0
		self.InvertedIndex = []

class Word_index:
	def __init__(self):
		self.word = ''
		self.index = []

class Paper:
	def __init__(self):
		self.identity = ''
		self.title = ''
		self.authors = []
		self.venue = Venue()
		self.year = 0
		self.keywords = []
		self.fos = []
		self.references = []
		self.n_citation = 0
		self.page_start = ''
		self.page_end = ''
		self.doc_type = ''
		self.lang = ''
		self.publisher = ''
		self.volume = ''
		self.issue = ''
		self.issn = ''
		self.isbn = ''
		self.doi = ''
		self.pdf = ''
		self.url = []
		self.abstract = ''
		self.indexed_abstract = Indexed_abstract()

class Collection:
	def __init__(self):
		self.collection_size = no_of_sample
		self.paper = []
		self.attributes = {}

	def print_data(self):
		for i in self.paper:
			print('identity : ' + i.identity)
			print('title : ' + i.title)
			print('authors :')
			for j in i.authors:
				print('\tname : ' + j.name + '\tidentity : ' + j.identity + '\torg : ' + j.org)
			print('keywords :')
			for j in i.keywords:
				print('\t' + j)
			print('fos :')
			for j in i.fos:
				print('\tname : ' + j.name + '\tw : ' + str(j.w))
			print('references :')
			for j in i.references:
				print('\t' + j)
			print('venue :')
			print('\traw : ' + i.venue.raw + '\tidentity : ' + i.venue.identity)
			print('year : ' + str(i.year))
			print('n_citation : ' + str(i.n_citation))
			print('page_start : ' + i.page_start)
			print('page_end : ' + i.page_end)
			print('doc_type : ' + i.doc_type)
			print('lang : ' + i.lang)
			print('publisher : ' + i.publisher)
			print('volume : ' + i.volume)
			print('issue : ' + i.issue)
			print('issn : ' + i.issn)
			print('isbn : ' + i.isbn)
			print('doi : ' + i.doi)
			print('pdf : ' + i.pdf)
			print('url :')
			for j in i.url:
				print('\t' + j)
			print('indexed_abstract :')
			print('\tIndexLength : ' + str(i.indexed_abstract.IndexLength))
			print('\tInvertedIndex : ')
			for j in i.indexed_abstract.InvertedIndex:
				print('\t\tword : ' + j.word, end = '')
				print('\t\tIndex : ', end = '')
				for k in j.index:
					print(str(k), end = ',')
				print()
			print('\n')

	def read_data(self):
		global j
		input_file = open(data_file_name, "r")
		for i in range(0, self.collection_size):
			this_paper = input_file.readline()
			#print(i)
			#print(this_paper + '\n')
			paper = Paper()
			j = 0;
			if(this_paper == ''):
				self.collection_size = i;
				break
			while j < len(this_paper):
				this_word = get_next_word(this_paper)
				if this_word in self.attributes:
					key = self.attributes[this_word]
					if(key == 0):
						paper.identity = get_next_word(this_paper)
					elif(key == 1):
						paper.title = get_next_word(this_paper)
					elif(key == 2):
						while j < len(this_paper) and this_paper[j + 1] != ']':
							temp = Authors()
							j = j + 1
							while j < len(this_paper) and this_paper[j] != '}':
								temp1 = get_next_word(this_paper)
								if(temp1 == 'name'):
									temp.name = get_next_word(this_paper)
								elif(temp1 == 'id'):
									temp.identity = get_next_word(this_paper)
								elif(temp1 == 'org'):
									temp.org = get_next_word(this_paper)
							paper.authors.append(temp)
					elif(key == 3):
						while j < len(this_paper) and this_paper[j] != '}':
							temp = get_next_word(this_paper)
							if(temp == 'id'):
								paper.venue.identity = get_next_word(this_paper)
							elif(temp == 'raw'):
								paper.venue.raw = get_next_word(this_paper)
					elif(key == 4):
						j = j + 2
						temp = ''
						while j < len(this_paper) and this_paper[j] >= '0' and this_paper[j] <= '9':
							temp = temp + this_paper[j]
							j = j + 1
						paper.year = int(temp)
					elif(key == 5):
						while j < len(this_paper) and this_paper[j] != ']':
							paper.keywords.append(get_next_word(this_paper))
					elif(key == 6):
						while j < len(this_paper) and this_paper[j + 1] != ']':
							temp = Fos()
							j = j + 1
							while j < len(this_paper) and this_paper[j] != '}':
								temp1 = get_next_word(this_paper)
								if(temp1 == 'name'):
									temp.name = get_next_word(this_paper)
								elif(temp1 == 'w'):
									j = j + 2
									temp2 = ''
									while j < len(this_paper) and ((this_paper[j] >= '0' and this_paper[j] <= '9') or this_paper[j] == '.'):
										temp2 = temp2 + this_paper[j]
										j = j + 1
									temp.w = float(temp2)
							paper.fos.append(temp)
					elif(key == 7):
						while j < len(this_paper) and this_paper[j] != ']':
							paper.references.append(get_next_word(this_paper))
					elif(key == 8):
						j = j + 2
						temp = ''
						while j < len(this_paper) and this_paper[j] >= '0' and this_paper[j] <= '9':
							temp = temp + this_paper[j]
							j = j + 1
						paper.n_citation = int(temp)
					elif(key == 9):
						paper.page_start = get_next_word(this_paper)
					elif(key == 10):
						paper.page_end = get_next_word(this_paper)
					elif(key == 11):
						paper.doc_type = get_next_word(this_paper)
					elif(key == 12):
						paper.lang = get_next_word(this_paper)
					elif(key == 13):
						paper.publisher = get_next_word(this_paper)
					elif(key == 14):
						paper.volume = get_next_word(this_paper)
					elif(key == 15):
						paper.issue = get_next_word(this_paper)
					elif(key == 16):
						paper.issn = get_next_word(this_paper)
					elif(key == 17):
						paper.isbn = get_next_word(this_paper)
					elif(key == 18):
						paper.doi = get_next_word(this_paper)
					elif(key == 19):
						paper.pdf = get_next_word(this_paper)
					elif(key == 20):
						while j < len(this_paper) and this_paper[j] != ']':
							paper.url.append(get_next_word(this_paper))
					elif(key == 21):
						paper.abstract = get_next_word(this_paper)
					elif(key == 22):
						get_next_word(this_paper)
						j = j + 2
						temp = ''
						while j < len(this_paper) and this_paper[j] >= '0' and this_paper[j] <= '9':
							temp = temp + this_paper[j]
							j = j + 1
						paper.indexed_abstract.IndexLength = int(temp)
						get_next_word(this_paper)
						while j < len(this_paper) and this_paper[j + 1] != '}':
							temp1 = Word_index()
							temp1.word = get_next_word(this_paper)
							j = j + 3
							temp2 = ''
							while j < len(this_paper) and this_paper[j] != ']':
								if(this_paper[j] == ','):
									temp1.index.append(temp2)
									temp2 = ''
									j = j + 1
								else:
									temp2 = temp2 + this_paper[j]
								j = j + 1
							temp1.index.append(temp2)
							paper.indexed_abstract.InvertedIndex.append(temp1)
			self.paper.append(paper)
		input_file.close()

	def map_attributes_to_interger(self):
		self.attributes['id'] = 0
		self.attributes['title'] = 1
		self.attributes['authors'] = 2
		self.attributes['venue'] = 3
		self.attributes['year'] = 4
		self.attributes['keywords'] = 5
		self.attributes['fos'] = 6
		self.attributes['references'] = 7
		self.attributes['n_citation'] = 8
		self.attributes['page_start'] = 9
		self.attributes['page_end'] = 10
		self.attributes['doc_type'] = 11
		self.attributes['lang'] = 12
		self.attributes['publisher'] = 13
		self.attributes['volume'] = 14
		self.attributes['issue'] = 15
		self.attributes['issn'] = 16
		self.attributes['isbn'] = 17
		self.attributes['doi'] = 18
		self.attributes['pdf'] = 19
		self.attributes['url'] = 20
		self.attributes['abstract'] = 21
		self.attributes['indexed_abstract'] = 22

collection = Collection()
collection.map_attributes_to_interger()
collection.read_data()
#collection.print_data()